File: src/main/java/org/example/expert/domain/manager/service/ManagerService.java
package org.example.expert.domain.manager.service;

import lombok.RequiredArgsConstructor;
import org.example.expert.exception.InvalidRequestException;
import org.example.expert.domain.manager.dto.request.ManagerSaveRequest;
import org.example.expert.domain.manager.dto.response.ManagerResponse;
import org.example.expert.domain.manager.dto.response.ManagerSaveResponse;
import org.example.expert.domain.manager.entity.Manager;
import org.example.expert.domain.manager.repository.ManagerRepository;
import org.example.expert.domain.todo.entity.Todo;
import org.example.expert.domain.todo.service.TodoService;
import org.example.expert.domain.user.dto.response.UserResponse;
import org.example.expert.domain.user.entity.User;
import org.example.expert.domain.user.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ManagerService {

    private final ManagerRepository managerRepository;
    private final UserRepository userRepository;
    private final TodoService todoService;

    @Transactional
    public ManagerSaveResponse saveManager(User user, Long todoId, ManagerSaveRequest managerSaveRequest) {
        User targetUser = userRepository.findById(managerSaveRequest.getTargetUserId())
                .orElseThrow(() -> new InvalidRequestException("등록하려고 하는 유저가 존재하지 않습니다"));
        Todo todo = todoService.findByIdOrFail(todoId);
        todo.isOwner(user);
        validateManagerAssignment(targetUser, todo);
        Manager manager = managerRepository.save(new Manager(targetUser, todo));

        return new ManagerSaveResponse(manager);
    }


    private User validateManagerAssignment(User targetUser, Todo todo){
        //이미 배정된 유저인지 확인
        if (managerRepository.findByUserAndTodo(targetUser, todo).isPresent()) {
            throw new InvalidRequestException("이미 해당 일정에 담당된 유저입니다");
        }
        todo.validateManagerAssignment(targetUser);
        return targetUser;
    }


    public List<ManagerResponse> getManagers(Long todoId) {
        Todo todo = todoService.findByIdOrFail(todoId);
        return managerRepository.findByTodoIdWithUser(todo.getId())
                .stream()
                .map(manager -> new ManagerResponse(manager.getId(), new UserResponse(manager.getUser())))
                .toList();
    }


    @Transactional
    public void deleteManager(User user, Long todoId, Long managerId) {
        Todo todo = todoService.findByIdOrFail(todoId);
        todo.isOwner(user);

        Manager manager = managerRepository.findById(managerId)
                .orElseThrow(() -> new InvalidRequestException("Manager not found"));

        manager.isAssigned(todoId);
        managerRepository.delete(manager);
    }
}

File: src/main/java/org/example/expert/service/ManagerService.kt
package org.example.expert.service

import org.example.expert.controller.manager.dto.request.ManagerCreateReqDto
import org.example.expert.controller.manager.dto.response.ManagerCreateRespDto
import org.example.expert.controller.manager.dto.response.ManagerInfoRespDto
import org.example.expert.domain.manager.Manager
import org.example.expert.domain.manager.ManagerRepository
import org.example.expert.domain.todo.Todo
import org.example.expert.domain.user.User
import org.example.expert.domain.user.UserRepository
import org.example.expert.exception.CustomApiException
import org.example.expert.exception.ErrorCode
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import kotlin.jvm.optionals.getOrElse

@Service
@Transactional(readOnly = true)
class ManagerService(
    private val userRepository: UserRepository,
    private val todoService: TodoService,
    private val managerRepository: ManagerRepository
) {

    @Transactional
    fun saveManager(user: User, todoId: Long, managerCreateReqDto: ManagerCreateReqDto): ManagerCreateRespDto {
        val targetUser = userRepository.findById(managerCreateReqDto.targetUserId)
            .getOrElse { throw CustomApiException(ErrorCode.TARGET_USER_NOT_FOUND) }

        val todo = todoService.findByIdOrFail(todoId)
        todo.isOwner(user)
        validateManagerAssignment(targetUser, todo)

        return ManagerCreateRespDto(
            managerRepository.save(managerCreateReqDto.toEntity(targetUser, todo))
        )
    }

    private fun validateManagerAssignment(targetUser: User, todo: Todo) {
        managerRepository.findByUserAndTodo(targetUser, todo)?.let {
            throw CustomApiException(ErrorCode.ALREADY_ASSIGNED_USER)
        }
        todo.validateManagerAssignment(targetUser)
    }

    fun getManagerList(todoId: Long): List<ManagerInfoRespDto> {
        val todo = todoService.findByIdOrFail(todoId)
        return managerRepository.findByTodoWithUser(todo)
            .map(::ManagerInfoRespDto)
    }

    @Transactional
    fun deleteManager(user: User, todoId: Long, managerId: Long){
        val todo = todoService.findByIdOrFail(todoId)
        todo.isOwner(user)

        val manager = findByIdOrFail(managerId)
        manager.isAssigned(todoId)

        managerRepository.delete(manager)
    }

    fun findByIdOrFail(managerId: Long): Manager = managerRepository
        .findById(managerId)
        .getOrElse{ throw CustomApiException(ErrorCode.MANAGER_NOT_FOUND) }
}